File: src/i18n.ts
/**
 * 軽量な多言語対応。
 * 現状はフロントUI（設定モーダル、ツールバーtooltip）のみカバーする。
 * メニューバー（Rust側）は別途翻訳が必要。
 */

export type Lang = "ja" | "en";

const DICT: Record<Lang, Record<string, string>> = {
  ja: {
    // 設定モーダル
    "settings.title": "設定",
    "settings.section.font": "フォント",
    "settings.section.display": "表示",
    "settings.section.language": "言語",
    "settings.display.theme": "テーマ",
    "settings.display.theme.system": "システムに合わせる",
    "settings.display.theme.dark": "ダーク",
    "settings.display.theme.light": "ライト",
    "settings.font.family": "フォント",
    "settings.font.codeFamily": "コード用フォント",
    "settings.font.codeColor": "コード文字色",
    "settings.font.codeColorFollow": "本文色に合わせる",
    "settings.font.size": "文字サイズ (px)",
    "settings.font.preview": "プレビュー: あいうえお ABCDE 12345",
    "settings.font.codePreview": "プレビュー: const x = 42;",
    "settings.font.custom": "（カスタム）{value}",
    // フォントプリセット
    "font.preset.systemDefault": "システム既定",
    "font.preset.yuGothicUI": "Yu Gothic UI（游ゴシック）",
    "font.preset.meiryo": "Meiryo（メイリオ）",
    "font.preset.msPGothic": "MS P ゴシック",
    "font.preset.notoSansJP": "Noto Sans JP",
    "font.preset.yuMincho": "Yu Mincho（游明朝）",
    "font.preset.msPMincho": "MS P 明朝",
    "font.preset.cascadiaCode": "Cascadia Code（等幅）",
    "font.preset.consolas": "Consolas（等幅）",
    "settings.display.showRecent": "最近使ったファイルをメニューに表示",
    "settings.language.select": "表示言語",
    "settings.language.system": "システムに合わせる",
    "settings.language.note":
      "OSのファイルダイアログ等、OSの言語設定に従う一部のテキストは切り替えられない場合があります。",
    // 確認ダイアログ
    "dlg.save.title": "保存しますか？",
    "dlg.save.body": "{filename} には未保存の変更があります。",
    "dlg.save.save": "保存",
    "dlg.save.discard": "破棄",
    "dlg.save.cancel": "キャンセル",
    "dlg.dup.title": "すでに開いています",
    "dlg.dup.body": "{filename} はすでに開かれています。どうしますか？",
    "dlg.dup.switch": "そのタブに切替",
    "dlg.dup.reload": "破棄して開き直す",
    "dlg.dup.cancel": "キャンセル",
    "dlg.closeAll.title": "未保存の変更があります",
    "dlg.closeAll.body": "未保存のタブがあります。どうしますか？",
    "dlg.closeAll.review": "確認する",
    "dlg.closeAll.discard": "破棄して終了",
    "dlg.closeAll.cancel": "キャンセル",
    "dlg.image.url": "画像のURLまたはパスを入力してください",
    "dlg.image.alt": "代替テキスト / キャプション（任意）",
    "dlg.image.size": "画像幅 (px) — 元サイズ: {natural}px / 空欄で自動",
    "dlg.image.sizeNoHint": "画像幅 (px、空欄で自動)",
    "settings.button.cancel": "キャンセル",
    "settings.button.apply": "適用",
    "settings.button.reset": "デフォルトに戻す",
    // ツールバー tooltip
    "tb.file_new": "新規 (Ctrl+N)",
    "tb.file_open": "開く (Ctrl+O)",
    "tb.file_save": "保存 (Ctrl+S)",
    "tb.file_save_as": "名前を付けて保存 (Ctrl+Shift+S)",
    "tb.h1": "見出し1",
    "tb.h2": "見出し2",
    "tb.h3": "見出し3",
    "tb.h4": "見出し4",
    "tb.bold": "太字 (Ctrl+B)",
    "tb.italic": "斜体 (Ctrl+I)",
    "tb.strike": "取り消し線 (Ctrl+Shift+X)",
    "tb.code": "インラインコード (Ctrl+E)",
    "tb.bullet": "箇条書きリスト",
    "tb.ordered": "番号付きリスト",
    "tb.quote": "引用",
    "tb.codeblock": "コードブロック",
    "tb.table": "表を挿入",
    "tb.link": "リンク (Ctrl+K)",
    "tb.image": "画像",
    "tb.hr": "区切り線",
    "tb.settings": "設定",
    "settings.button.preview": "プレビュー",
  },
  en: {
    "settings.title": "Settings",
    "settings.section.font": "Font",
    "settings.section.display": "Display",
    "settings.section.language": "Language",
    "settings.display.theme": "Theme",
    "settings.display.theme.system": "Match system",
    "settings.display.theme.dark": "Dark",
    "settings.display.theme.light": "Light",
    "settings.font.family": "Font family",
    "settings.font.codeFamily": "Code font family",
    "settings.font.codeColor": "Code text color",
    "settings.font.codeColorFollow": "Match body text color",
    "settings.font.size": "Font size (px)",
    "settings.font.preview": "Preview: The quick brown fox 12345",
    "settings.font.codePreview": "Preview: const x = 42;",
    "settings.font.custom": "(custom) {value}",
    "font.preset.systemDefault": "System default",
    "font.preset.yuGothicUI": "Yu Gothic UI",
    "font.preset.meiryo": "Meiryo",
    "font.preset.msPGothic": "MS P Gothic",
    "font.preset.notoSansJP": "Noto Sans JP",
    "font.preset.yuMincho": "Yu Mincho",
    "font.preset.msPMincho": "MS P Mincho",
    "font.preset.cascadiaCode": "Cascadia Code (monospace)",
    "font.preset.consolas": "Consolas (monospace)",
    "settings.display.showRecent": "Show recent files in menu",
    "settings.language.select": "Language",
    "settings.language.system": "Match system",
    "settings.language.note":
      "Some texts (e.g. native OS file dialogs) follow the OS language and may not switch.",
    "dlg.save.title": "Save changes?",
    "dlg.save.body": "{filename} has unsaved changes.",
    "dlg.save.save": "Save",
    "dlg.save.discard": "Discard",
    "dlg.save.cancel": "Cancel",
    "dlg.dup.title": "Already open",
    "dlg.dup.body": "{filename} is already open. What do you want to do?",
    "dlg.dup.switch": "Switch to tab",
    "dlg.dup.reload": "Discard and reopen",
    "dlg.dup.cancel": "Cancel",
    "dlg.closeAll.title": "Unsaved changes",
    "dlg.closeAll.body": "There are unsaved tabs. What do you want to do?",
    "dlg.closeAll.review": "Review",
    "dlg.closeAll.discard": "Discard and quit",
    "dlg.closeAll.cancel": "Cancel",
    "dlg.image.url": "Enter image URL or path",
    "dlg.image.alt": "Alt text / caption (optional)",
    "dlg.image.size": "Image width (px) — original: {natural}px / leave empty for auto",
    "dlg.image.sizeNoHint": "Image width (px, leave empty for auto)",
    "settings.button.cancel": "Cancel",
    "settings.button.apply": "Apply",
    "settings.button.reset": "Reset to defaults",
    "tb.file_new": "New (Ctrl+N)",
    "tb.file_open": "Open (Ctrl+O)",
    "tb.file_save": "Save (Ctrl+S)",
    "tb.file_save_as": "Save as (Ctrl+Shift+S)",
    "tb.h1": "Heading 1",
    "tb.h2": "Heading 2",
    "tb.h3": "Heading 3",
    "tb.h4": "Heading 4",
    "tb.bold": "Bold (Ctrl+B)",
    "tb.italic": "Italic (Ctrl+I)",
    "tb.strike": "Strikethrough (Ctrl+Shift+X)",
    "tb.code": "Inline code (Ctrl+E)",
    "tb.bullet": "Bulleted list",
    "tb.ordered": "Numbered list",
    "tb.quote": "Quote",
    "tb.codeblock": "Code block",
    "tb.table": "Insert table",
    "tb.link": "Link (Ctrl+K)",
    "tb.image": "Image",
    "tb.hr": "Horizontal rule",
    "tb.settings": "Settings",
    "settings.button.preview": "Preview",
  },
};

export const LANG_OPTIONS: { value: Lang; label: string }[] = [
  { value: "ja", label: "日本語" },
  { value: "en", label: "English" },
];

let currentLang: Lang = "ja";
const listeners = new Set<(lang: Lang) => void>();

export function setLang(lang: Lang): void {
  if (lang === currentLang) return;
  currentLang = lang;
  for (const fn of listeners) fn(lang);
}

export function getLang(): Lang {
  return currentLang;
}

export function t(key: string): string {
  return DICT[currentLang][key] ?? DICT.ja[key] ?? key;
}

export function onLangChange(fn: (lang: Lang) => void): () => void {
  listeners.add(fn);
  return () => listeners.delete(fn);
}

File: src-tauri/src/i18n.rs
use std::sync::Mutex;
use tauri::{AppHandle, Manager, State};

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Lang {
    Ja,
    En,
}

impl Lang {
    pub fn from_code(s: &str) -> Self {
        match s {
            "en" | "EN" | "en-US" => Lang::En,
            _ => Lang::Ja,
        }
    }
}

pub struct LangState(pub Mutex<Lang>);

pub fn current(app: &AppHandle) -> Lang {
    let state: State<LangState> = app.state();
    let guard = state.0.lock();
    match guard {
        Ok(g) => *g,
        Err(_) => Lang::Ja,
    }
}

pub fn set(app: &AppHandle, lang: Lang) {
    let state: State<LangState> = app.state();
    let guard = state.0.lock();
    if let Ok(mut g) = guard {
        *g = lang;
    }
}

pub fn t(lang: Lang, key: &str) -> String {
    use Lang::*;
    let s: &str = match (lang, key) {
        // 上位メニュー
        (Ja, "menu.file") => "ファイル",
        (En, "menu.file") => "File",
        (Ja, "menu.edit") => "編集",
        (En, "menu.edit") => "Edit",
        (Ja, "menu.format") => "書式",
        (En, "menu.format") => "Format",
        (Ja, "menu.view") => "表示",
        (En, "menu.view") => "View",
        (Ja, "menu.help") => "ヘルプ",
        (En, "menu.help") => "Help",

        // ファイル
        (Ja, "file.new") => "新規タブ",
        (En, "file.new") => "New tab",
        (Ja, "file.open") => "開く...",
        (En, "file.open") => "Open...",
        (Ja, "file.save") => "保存",
        (En, "file.save") => "Save",
        (Ja, "file.save_as") => "名前を付けて保存...",
        (En, "file.save_as") => "Save as...",
        (Ja, "file.close") => "タブを閉じる",
        (En, "file.close") => "Close tab",
        (Ja, "file.history_more") => "ファイルヒストリの続き",
        (En, "file.history_more") => "More recent files",
        (Ja, "file.quit") => "終了",
        (En, "file.quit") => "Quit",

        // 書式
        (Ja, "fmt.bold") => "太字",
        (En, "fmt.bold") => "Bold",
        (Ja, "fmt.italic") => "斜体",
        (En, "fmt.italic") => "Italic",
        (Ja, "fmt.strike") => "取り消し線",
        (En, "fmt.strike") => "Strikethrough",
        (Ja, "fmt.code") => "インラインコード",
        (En, "fmt.code") => "Inline code",
        (Ja, "fmt.h1") => "見出し1",
        (En, "fmt.h1") => "Heading 1",
        (Ja, "fmt.h2") => "見出し2",
        (En, "fmt.h2") => "Heading 2",
        (Ja, "fmt.h3") => "見出し3",
        (En, "fmt.h3") => "Heading 3",
        (Ja, "fmt.quote") => "引用",
        (En, "fmt.quote") => "Quote",
        (Ja, "fmt.bullet") => "箇条書きリスト",
        (En, "fmt.bullet") => "Bulleted list",
        (Ja, "fmt.ordered") => "番号付きリスト",
        (En, "fmt.ordered") => "Numbered list",
        (Ja, "fmt.codeblock") => "コードブロック",
        (En, "fmt.codeblock") => "Code block",
        (Ja, "fmt.table") => "表を挿入",
        (En, "fmt.table") => "Insert table",
        (Ja, "fmt.hr") => "区切り線",
        (En, "fmt.hr") => "Horizontal rule",
        (Ja, "fmt.link") => "リンク",
        (En, "fmt.link") => "Link",

        // 表示
        (Ja, "view.zoom_in") => "文字サイズを大きく",
        (En, "view.zoom_in") => "Zoom in",
        (Ja, "view.zoom_out") => "文字サイズを小さく",
        (En, "view.zoom_out") => "Zoom out",
        (Ja, "view.zoom_reset") => "文字サイズをリセット",
        (En, "view.zoom_reset") => "Reset zoom",
        (Ja, "view.settings") => "設定...",
        (En, "view.settings") => "Settings...",

        // ヘルプ
        (Ja, "help.about") => "mdeditについて",
        (En, "help.about") => "About mdedit",

        _ => key,
    };
    s.to_string()
}
